### PR TITLE
[Async TP] Fix dim swapping before reduction in fused_scaled_matmul_reduce_scatter

### PR DESCRIPTION
**Pull request by danielvegamyhre**
## Summary
- The unit test `pytest test/distributed/test_symmetric_memory.py -k test_fused_scaled_matmul_reduce_scatter_scatter` was not running for some reason when #149247 was merged, giving false green CI signals. When it was ran manually recently, the test failed, highlighting a bug causing incorrect numerics when `scatter_dim=1`.
- This PR fixes the bug, which was related to how we swap dims 0<=>scatter_dim at the beginning of the custom op (for more efficient cross-device data movement I believe), then swap it back prior to reduction. 

## Test plan
- I confirmed the unit test `pytest test/distributed/test_symmetric_memory.py -k test_fused_scaled_matmul_reduce_scatter_scatter` is now passing.
- I confirmed e2e training w/ torchtitan looks good ([logs](https://www.internalfb.com/phabricator/paste/view/P1812054188))
- I analyzed the tlparse to verify the fused_all_gather_matmul and fused_scaled_matmul_reduce_scatter both appear at least once in the post grad graphs ([tlparse](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpVbUsdG/dedicated_log_torch_trace_65oh3qj_.log/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000))

## Next steps
1. I think for async TP `fused_scaled_matmul_reduce_scatter` we may only need `scatter_dim_after_maybe_reshape` and not `orig_scatter_dim` after all. I can confirm this and refactor if it is the case.
2. This op is specifically designed for async TP, and many of the arguments don't make sense for a user trying to use this as a standalone op. IMO we should have separate standalone custom op without all the extra function args and internal logic that doesn't apply to non-async TP cases. 
3. In a follow up PR I want to add shape annotations to each line (e.g. `# (B, T, H)` etc) to make this easier to debug in the future.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
cc @fegin @tianyu-l @kwen2501

**Review by fegin — APPROVED** [review]
Thanks for fixing the issue. I'm curious that does this mean that there should be some loss issue without this fix?

**Comment by danielvegamyhre** [comment]
> Thanks for fixing the issue. I'm curious that does this mean that there should be some loss issue without this fix?

The loss drops more quickly down with the fix. Without the fix it seems to slow down around 7.5, but with the fix it is still dropping at step 100 w/ loss ~7.0. See logs without the fix: https://www.internalfb.com/phabricator/paste/view/P1812453218 (compare this with logs with the fix linked in PR description).

This must be a recent regression though, perhaps due to an interaction with https://github.com/pytorch/pytorch/pull/150935, because in the original PR (https://github.com/pytorch/pytorch/pull/149247) you can see in the loss curve screenshots I included that it is identical to bf16 through step 100 and hits ~7.0 after 100 steps.
